src/test/mod.rs

```diff
@@ -2,13 +2,55 @@ mod fakes;
 
 use crate::web::Server;
 use failure::Error;
+use once_cell::unsync::OnceCell;
 use postgres::Connection;
-use std::sync::{Arc, Mutex, MutexGuard};
 use reqwest::{Client, Method, RequestBuilder};
+use std::sync::{Arc, Mutex, MutexGuard};
+
+pub(crate) fn wrapper(f: impl FnOnce(&TestEnvironment) -> Result<(), Error>) {
+    let env = TestEnvironment::new();
+    let result = f(&env);
+    env.cleanup();
+
+    if let Err(err) = result {
+        eprintln!("the test failed: {}", err);
+        for cause in err.iter_causes() {
+            eprintln!("  caused by: {}", cause);
+        }
+
+        eprintln!("{}", err.backtrace());
 
-pub(crate) fn with_database(f: impl FnOnce(&TestDatabase) -> Result<(), Error>) {
-    let env = TestDatabase::new().expect("failed to initialize the environment");
-    report_error(|| f(&env));
+        panic!("the test failed");
+    }
+}
+
+pub(crate) struct TestEnvironment {
+    db: OnceCell<TestDatabase>,
+    frontend: OnceCell<TestFrontend>,
+}
+
+impl TestEnvironment {
+    fn new() -> Self {
+        Self {
+            db: OnceCell::new(),
+            frontend: OnceCell::new(),
+        }
+    }
+
+    fn cleanup(self) {
+        if let Some(frontend) = self.frontend.into_inner() {
+            frontend.server.leak();
+        }
+    }
+
+    pub(crate) fn db(&self) -> &TestDatabase {
+        self.db
+            .get_or_init(|| TestDatabase::new().expect("failed to initialize the db"))
```

Comment:
Do we want the tests to see each others changes? I'm worried that will lead to non-deterministic tests, especially since `cargo test` runs tests in parallel in an arbitrary order.

Reply:
Tests will not see each other changes: each test has its own instace of `TestEnvironment`, and the `OnceCell`s are fields inside it. That's the reason I added the `once_cell` crate: compared to lazy_static you can make non-global lazy fields.

Reply:
I missed that, that makes a lot of sense!